### ISSUE TITLE
bug(datepicker): Range picker fails when the adapter uses classes that can't be converted to primitives

### ISSUE THREAD

**Issue by robmv** [angular/components#27255]
### Is this a regression?

- [ ] Yes, this behavior used to work in the previous version

### Description

When using a custom DateAdapter for a date class that doesn't support conversion to primitives (js-joda), that I like because it avoids too many errors. It fails when a date is selected on the date range picker, with the following error:

_Uncaught TypeError: A conversion from Temporal to a number is not allowed. To compare use the methods .equals(), .compareTo(), .isBefore() or one that is more suitable to your use case._

This happen because  the code is using `cell.rawValue` at the following line:

https://github.com/angular/components/blob/1da7f15ff66ce30ad9888d56760bb149c960d2a6/src/material/datepicker/calendar-body.ts#LL515C40-L515C48

I think it should be using `cell.compareValue` instead.

A workaround can be made monkey patching `MatCalendarBody` (remember to use your date class check if not using js-joda)

```typescript
import { MatCalendarBody } from '@angular/material/datepicker';

const originalIsInRange = MatCalendarBody.prototype._isInRange;
MatCalendarBody.prototype._isInRange = function(value: number) {
  const badValue = value as any;
  if (badValue instanceof LocalDate) {
    value = new Date(badValue.year(), badValue.monthValue(), badValue.dayOfMonth()).getTime();
  }
  return originalIsInRange.call(this, value);
};
```

### Reproduction

I am pointing to the exact source line with the problem

### Expected Behavior

Using a custom DateAdapter for js-joda should work. The date range picker should not give an error with clicked.

### Actual Behavior

An exception when a date is clicked and the date class using by the DateAdapter doesn't support conversion to primitives.

### Environment

Angular CLI: 16.0.3
Node: 18.16.0
Package Manager: npm 9.5.1
OS: linux x64

Angular: 16.0.3
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, localize, platform-browser
... platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1600.3
@angular-devkit/build-angular   16.0.3
@angular-devkit/core            16.0.3
@angular-devkit/schematics      16.0.3
@angular/cdk                    16.0.2
@angular/material               16.0.2
@schematics/angular             16.0.3
ng-packagr                      16.0.1
rxjs                            7.8.1
typescript                      5.0.4

**Comment by zarend** [comment]
Hello,

Thank you for investigating this.

@crisbeto is there a specific reason why we are comparing values with `<=`/`>=` rather than through the date adapter API? Is it done that way for performance reasons. Is looks to me like doing this with the date adapter API (`compareDate`) could be more robust.

-Zach

**Comment by robmv** [comment]
I nearly forgot about this. Can this be fixed with the change we propose here or it must wait for the review about the need of this kind of comparison asked by @zarend?

**Comment by robmv** [comment]
This looks fixed on the main and 17.0.x branch

https://github.com/angular/components/blob/main/src/material/datepicker/calendar-body.ts#L518
https://github.com/angular/components/blob/17.0.x/src/material/datepicker/calendar-body.ts#L518

Not on 17.0.0, but it should be on a next fix release I think.

https://github.com/angular/components/blob/17.0.0/src/material/datepicker/calendar-body.ts#L515
